pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -226,6 +244,58 @@ func (h *HandlerImpl) Allocate(pod *v1.Pod, container *v1.Container, activePods
 	return h.writeCheckpoint()
 }
 
+// Allocate attempts to allocate all of required device plugin resources,
+// and update Allocatable resources in nodeInfo if necessary
+func (h *HandlerImpl) Allocate(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
+	pod := attrs.Pod
+	// TODO: Reuse devices between init containers and regular containers.
+	for _, container := range pod.Spec.InitContainers {
+		if err := h.allocateContainerResources(pod, &container); err != nil {
+			return err
+		}
+	}
+	for _, container := range pod.Spec.Containers {
+		if err := h.allocateContainerResources(pod, &container); err != nil {
+			return err
+		}
+	}
+
+	// quick return if no pluginResources requested
+	if _, podRequireDevicePluginResource := h.podDevices[string(pod.UID)]; !podRequireDevicePluginResource {
+		return nil
+	}
+
+	h.sanitizeNodeAllocatable(node)
+
+	return nil
+}
+
+// sanitizeNodeAllocatable scans through allocatedDevices in DevicePluginHandler,
+// and update allocatableResource in nodeInfo if necessary
+func (h *HandlerImpl) sanitizeNodeAllocatable(node *schedulercache.NodeInfo) {
```

Comment:
I'm a bit confused by the scenario of why this is needed?

From what I understand when the node advertises more devices than needed, we set allocatable to the number of devices needed right ?
I'm probably missing the bigger picture filters that get applied after this.

Would be helpful to add a comment explaining the logic of why this is needed :)

Reply:
@RenaudWasTaken The issues mentioned in this PR described the problem and why we need this change. PTAL.

Reply:
> For example, a device may have just turned unhealthy and the scheduler may not have noticed it. Note there can be multiple entities scheduling pods on to nodes.

That's not really what I'm confused by.
I'm confused by the function logic. From what I understand:
- it loops over the cached allocated devices 
  - if the node advertises less devices than needed it does nothing
    - Seems logical up to this point
  - if the node advertises more devices then are needed
    - Clones the allocatable field
    - Sets the allocatable field to the number of devices allocated
       - This is where I find this very confusing
       - I'd expect this to be set to the number of "healthy" devices not the number of allocated devices
       - This behavior looks like it might lead to subtle bugs

Anyways, a comment about what that function is trying to achieve seems like a good idea :)

Reply:
I'd expect the logic to be based on healthy devices rather than allocated devices which seems pretty counter intuitive

Reply:
@RenaudWasTaken You can take a look at this issue comment for the background information: https://github.com/kubernetes/kubernetes/issues/53395#issuecomment-336584099

@lichuqiang Could you also refer to that issue in your PR description? Maybe you can also update the comment above this function to be more specific, like "sanitizeNodeAllocatable scans through allocatedDevices in DevicePluginHandler and if necessary, updates allocatableResource in nodeInfo to at least equal to the allocated capacity. This allows pods that have already been scheduled on the node to pass GeneralPredicates admission checking even upon device plugin failure".

Reply:
Added. Yes, guys who don't know the background may easily get confused if no comment here.